Public/Src/Cache/ContentStore/Distributed/NuCache/RocksDbContentLocationDatabase.cs

```diff
@@ -507,6 +529,162 @@ private byte[] GetKey(ShortHash hash)
+                            store.Put(key, SerializeContentHashList(contentHashListWithDeterminism), nameof(Columns.CaChaaS));
+
+                            // Accept the value
+                            return new AddOrGetContentHashListResult(new ContentHashListWithDeterminism(null, determinism));
+                        }
+
+                        // If we didn't accept the new value because it is the same as before, just with a not
+                        // necessarily better determinism, then let the user know.
+                        if (oldContentHashList.Equals(contentHashList))
+                        {
+                            return new AddOrGetContentHashListResult(new ContentHashListWithDeterminism(null, oldDeterminism));
+                        }
+
+                        // If we didn't accept a deterministic tool's data, then we're in an inconsistent state
+                        if (determinism.IsDeterministicTool)
+                        {
+                            return new AddOrGetContentHashListResult(
+                                AddOrGetContentHashListResult.ResultCode.InvalidToolDeterminismError,
+                                oldContentHashListWithDeterminism.ContentHashListWithDeterminism);
+                        }
+
+                        // If we did not accept the given value, return the value in the cache
+                        return new AddOrGetContentHashListResult(oldContentHashListWithDeterminism);
+                    }
+
+                    return new AddOrGetContentHashListResult("Hit too many races attempting to add content hash list into the cache");
+                });
+
+            // Success for this status here may actually be an error that's been returned
+            return status.Succeeded ? status.Result : new AddOrGetContentHashListResult(status.Failure.CreateException());
+        }
+
+        private bool IsContentAvailable(OperationContext context, ContentHashList contentHashList)
+        {
+            // TODO(jubayard): implement. Since this needs to pin stuff, we need it to happen in a higher layer.
+            return true;
+        }
+
+        private byte[] SerializeContentHashList(ContentHashListWithDeterminism value)
+        {
+            return SerializeWithPool(value, (instance, writer) => instance.Serialize(writer));
+        }
+        
+        /// <inheritdoc />
+        public override IReadOnlyCollection<GetSelectorResult> GetSelectors(OperationContext context, Fingerprint weakFingerprint)
+        {
+            var result = new List<GetSelectorResult>();
+
+            var status = _keyValueStore.Use(
+                store =>
+                {
+                    var key = SerializeWeakFingerprint(weakFingerprint);
+
+                    // This only works because the strong fingerprint serializes the weak fingerprint first. Hence,

```

Comment:
Do we want to make sure this is always the case?

I can see a few options:
1. Add a test to make sure the behavior is enforced.
2. + change the design in a way that this important decision is explicit in the code and not "accidental".